runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+package org.corfudb.runtime;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterators;
+import lombok.Getter;
+import lombok.Setter;
+import org.corfudb.protocols.logprotocol.CheckpointEntry;
+import org.corfudb.protocols.logprotocol.SMREntry;
+import org.corfudb.runtime.collections.SMRMap;
+import org.corfudb.runtime.object.transactions.AbstractTransactionalContext;
+import org.corfudb.runtime.object.transactions.TransactionalContext;
+import org.corfudb.runtime.view.stream.BackpointerStreamView;
+import org.corfudb.util.serializer.Serializers;
+
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/** Checkpoint writer for SMRMaps: take a snapshot of the
+ *  object via TXBegin(), then dump the frozen object's
+ *  state into CheckpointEntry records into the object's
+ *  stream.
+ *
+ * TODO: Generalize to all SMR objects.
+ */
+public class CheckpointWriter {
+    /** Metadata to be stored in the CP's 'dict' map.
+     */
+    private UUID streamID;
+    private String author;
+    @Getter
+    private UUID checkpointID;
+    private LocalDateTime startTime;
+    private long startAddress, endAddress;
+    private long numEntries = 0, numBytes = 0;
+    Map<String, String> mdKV = new HashMap<>();
+
+    /** Mutator lambda to change map key.  Typically used for
+     *  testing but could also be used for type conversion, etc.
+     */
+    @Getter
+    @Setter
+    Function<Object,Object> keyMutator = (x) -> x;
+
+    /** Mutator lambda to change map value.  Typically used for
+     *  testing but could also be used for type conversion, etc.
+     */
+    @Getter
+    @Setter
+    Function<Object,Object> valueMutator = (x) -> x;
+
+    /** Batch size: number of SMREntry in a single CONTINUATION.
+     */
+    @Getter
+    @Setter
+    private int batchSize = 50;
+
+    /** BiConsumer to run after every CheckpointEntry is appended
+     * to the stream
+     */
+    @Getter
+    @Setter
+    BiConsumer<CheckpointEntry,Long> postAppendFunc = (cp, l) -> {};
+
+    /** Local ref to the object's runtime.
+     */
+    private CorfuRuntime rt;
+
+    /** Local ref to the stream's view.
+     */
+    BackpointerStreamView sv;
```

Comment:
any reason why you are not using the interface instead? also why not just use the runtime's streamsview?